index.d.ts

```diff
@@ -1,12 +1,65 @@
-import { Parser } from 'acorn' 
+import * as acorn from 'acorn';
 
-declare const jsx: (options?: jsx.Options) => (BaseParser: typeof Parser) => typeof Parser;
+declare const jsx: {
+  tokTypes: typeof acorn.tokTypes;
+  (options?: jsx.Options): (BaseParser: typeof acorn.Parser) => jsx.AcornJsxParser
+}
 
 declare namespace jsx {
-  interface Options {
+
+  type tokTypes = typeof acorn.tokTypes;
+
+  export interface TokTypes extends tokTypes {
+    jsxName: acorn.TokenType,
+    jsxText: acorn.TokenType,
+    jsxTagEnd: acorn.TokenType,
+    jsxTagStart: acorn.TokenType
+  }
+
+  export interface Options {
     allowNamespacedObjects?: boolean;
     allowNamespaces?: boolean;
   }
+
+  export type TokContexts = {
+    tc_oTag: acorn.TokContext,
+    tc_cTag: acorn.TokContext,
+    tc_expr: acorn.TokContext
+  };
+
+  type P = typeof acorn.Parser;
+
+  // We pick (statics) from acorn rather than plain extending to avoid complaint
+  //   about base constructors needing the same return type (i.e., we return
+  //   `IAcornJsxParser` here)
+  export interface AcornJsxParser extends Pick<P, keyof P> {
```

Comment:
I don’t think we need to expose the `P` type. It can be kept internal by moving it out of the namespace. Altough personally I prefer to avoid it at all. It’s only used twice.

Also constructor types are often postfixed with `Ctor`

```suggestion
  // We pick (statics) from acorn rather than plain extending to avoid complaint
  //   about base constructors needing the same return type (i.e., we return
  //   `AcornJsxParser` here)
  export interface AcornJsxParserCtor extends Pick<typeof acorn.Parser, keyof typeof acorn.Parser> {
```

Reply:
Ok, I've renamed to use `Ctor` and also inlined the type alias.